Basestation_Software.Web/Core/Components/PMS.razor

```diff
@@ -0,0 +1,186 @@
+        <button style="grid-area: reboot" onclick="@Reboot">REBOOT</button>
+        <button style="grid-area: e-stop" onclick="@EStop">E-STOP</button>
+        <button style="grid-area: suicide" onclick="@Suicide">SUICIDE</button>
+    </div>
+</div>
+
+@code
+{
+    enum BusState
+    {
+        DISABLED,
+        ENABLED,
+        UNKNOWN,
+    }
+
+    // Constants
+    public const uint TELEMETRY_TIMEOUT = 10_000; // ms
+    public static Color COLOR_ERROR = Color.FromArgb(255, 255, 0, 0);
+    public static Color COLOR_OK = Color.FromArgb(255, 0, 255, 0);
+
+    // State
+    private List<BusState> BusStatus = new List<BusState>() { BusState.UNKNOWN, BusState.UNKNOWN, BusState.UNKNOWN }; // Motors, Core, Aux
+    private float PackCurrent = 0, PackVoltage = 0, AuxCurrent = 0;
+    private List<float> CellVoltage = new List<float>() { 0, 0, 0, 0, 0, 0 }; // C1 ... C6
+    private List<float> MiscCurrent = new List<float>() { 0, 0, 0 }; // CS1 ... CS3
+    private System.Threading.Timer TelemetryWatchdog;
+
+    protected override void OnInitialized()
+    {
+        TelemetryWatchdog = new System.Threading.Timer(WatchdogTimeout, null, TELEMETRY_TIMEOUT, Timeout.Infinite);
+
+        _RoveCommService.On<float>("PMS", "PackCurrent", async (RoveCommPacket<float> packet) =>
+        {
+            PackCurrent = packet.Data[0];
+            ResetWatchdog();
+            await InvokeAsync(StateHasChanged);
+        });
+        _RoveCommService.On<float>("PMS", "PackVoltage", async (RoveCommPacket<float> packet) =>
+        {
+            PackVoltage = packet.Data[0];
+            ResetWatchdog();
+            await InvokeAsync(StateHasChanged);
+        });
+        _RoveCommService.On<float>("PMS", "CellVoltage", async (RoveCommPacket<float> packet) =>
+        {
+            CellVoltage = packet.Data;
+            ResetWatchdog();
+            await InvokeAsync(StateHasChanged);
+        });
+        _RoveCommService.On<float>("PMS", "AuxCurrent", async (RoveCommPacket<float> packet) =>
+        {
+            AuxCurrent = packet.Data[0];
+            ResetWatchdog();
+            await InvokeAsync(StateHasChanged);
+        });
+        _RoveCommService.On<float>("PMS", "MiscCurrent", async (RoveCommPacket<float> packet) =>
+        {
+            MiscCurrent = packet.Data;
+            ResetWatchdog();
+            await InvokeAsync(StateHasChanged);
+        });
+        _RoveCommService.On<float>("PMS", "PackVoltage", async (RoveCommPacket<float> packet) =>
+        {
+            PackVoltage = packet.Data[0];
+            ResetWatchdog();
+            await InvokeAsync(StateHasChanged);
+        });
+        _RoveCommService.On<byte>("PMS", "BusStatus", async (RoveCommPacket<byte> packet) =>
+        {
+            for (int i = 0; i < 3; i++)
+                BusStatus[i] = (packet.Data[0] & (1 << i)) == 0 ? BusState.DISABLED : BusState.ENABLED;
+            ResetWatchdog();
+            await InvokeAsync(StateHasChanged);
+        });
```

Comment:
I don't imagine RoveComm C# slowing down with lots of callbacks added, but if multiple clients connect and it does, it might benefit to have a PMSService that does this part and the components get the data from that. I doubt that will be necessary though, just putting this here for future note. Not having an extra service is preferred.

Reply:
Unless you have multiple PMS components on a single page, I expect performance with multiple clients to be the same with and without a PMS service. Rovecomm or a PMS service would still have to loop over and notify multiple clients.